src/control/zoom/ZoomControl.cpp

```diff
@@ -105,35 +113,64 @@ void ZoomControl::zoomScroll(ZoomDirection zoomIn, utl::Point<double> zoomCenter
         this->setZoomFitMode(false);
     }
 
-    if (this->zoomSequenceStart == -1 || scrollCursorPosition != zoomCenter) {
-        scrollCursorPosition = zoomCenter;
-        startZoomSequence(zoomCenter);
-    }
+    double multiplier = 1.0 + this->zoomStepScroll;
+    double newZoomStep = zoomIn ? multiplier : 1.0 / multiplier;
+    double newZoom = this->zoom * newZoomStep;
 
-    double newZoom = this->zoom + (zoomIn ? this->zoomStepScroll : -this->zoomStepScroll);
+    startZoomSequence(zoomCenter);
     this->zoomSequenceChange(newZoom, false);
+    endZoomSequence();
 }
 
 void ZoomControl::startZoomSequence() {
     Rectangle rect = getVisibleRect();
-    startZoomSequence({rect.x + rect.width / 2.0, rect.y + rect.height / 2.0});
+    startZoomSequence({rect.width / 2.0, rect.height / 2.0});
 }
 
 void ZoomControl::startZoomSequence(utl::Point<double> zoomCenter) {
+    // * set zoom center and zoom startlevel
+    this->zoomWidgetPos = zoomCenter;  // window space coordinates of the zoomCenter!
+    this->zoomSequenceStart = this->zoom;
+
+    // * set unscaledPixels padding value
+    size_t currentPageIdx = this->view->getCurrentPage();
+
+    // To get the layout, we need to call view->getWidget(), which isn't const.
+    // As such, we get the view and determine `unscaledPixels` here, rather than
+    // in `getScrollPositionAfterZoom`.
+    GtkWidget* widget = view->getWidget();
+    Layout* layout = gtk_xournal_get_layout(widget);
+
+    // Not everything changes size as we zoom in/out. The padding, for example,
+    // remains constant! (changed when page changes, but the error stays small enough)
+    this->unscaledPixels = {static_cast<double>(layout->getPaddingLeftOfPage(currentPageIdx)),
+                            static_cast<double>(layout->getPaddingAbovePage(currentPageIdx))};
+
+    // * set initial scrollPosition value
     auto const& rect = getVisibleRect();
     auto const& view_pos = utl::Point{rect.x, rect.y};
 
-    this->zoomWidgetPos = zoomCenter - view_pos;
-    this->zoomSequenceStart = this->zoom;
-
-    setScrollPositionAfterZoom(view_pos);
+    // Use this->zoomWidgetPos to zoom into a location other than the top-left (e.g. where
+    // the user pinched).
+    this->scrollPosition = (view_pos + this->zoomWidgetPos - this->unscaledPixels) / this->zoom;
 }
 
 void ZoomControl::zoomSequenceChange(double zoom, bool relative) {
-    if (relative && this->zoomSequenceStart != -1) {
-        zoom *= zoomSequenceStart;
+    if (relative) {
+        zoom *= this->zoomSequenceStart != -1 ? zoomSequenceStart : this->zoom;
```

Comment:
Please do not use inline if statements. They are hard to read and therefore error-prone.